### PR TITLE
GridTools::find_active_cell_around_point(): improve detection if point is outside of domain

### PR DESCRIPTION
**Pull request by peterrum**
FYI @mschreter

### CODE REVIEW

File: include/deal.II/grid/grid_tools.h [hunk 1]
```diff
@@ -1501,7 +1501,13 @@ namespace GridTools
     const std::vector<bool> &                              marked_vertices = {},
     const RTree<std::pair<Point<spacedim>, unsigned int>> &used_vertices_rtree =
       RTree<std::pair<Point<spacedim>, unsigned int>>{},
-    const double tolerance = 1.e-10);
+    const double tolerance = 1.e-10,
+    const RTree<
+      std::pair<BoundingBox<spacedim>,
+                typename Triangulation<dim, spacedim>::active_cell_iterator>>
+      &rtree = RTree<std::pair<
```
Comment (kronbichler) [inline include/deal.II/grid/grid_tools.h:1508]: Can you give this a more expressive name? See e.g. the `used_vertices_rtree` above.

**Review by kronbichler — APPROVED** [review]
Thanks!

**Comment by mathsen** [comment]
Hello,
this commit breaks compilation of deal.ii for me (using boost 1.76).
I don't know if the boost version is the culprit, the log is full with:
````
../source/grid/grid_tools.cc:2793:30: error: no matching function for call to 'boost::geometry::model::point<double, 1, boost::geometry::cs::cartesian>::point(const value_type&, const value_type&)'
2793 |             (spacedim == 2 ? point_t(p[0], p[1]) : point_t(p[0], p[1], p[2]));
[...]
/boost-1.76.0/include/boost/geometry/geometries/point.hpp:155:107: error: no type named 'type' in 'struct std::enable_if<false, int>'

[...]

../source/grid/grid_tools.cc:2793:52: error: no matching function for call to 'boost::geometry::model::point<double, 1, boost::geometry::cs::cartesian>::point(const value_type&, const value_type&, const value_type&)'
2793 |             (spacedim == 2 ? point_t(p[0], p[1]) : point_t(p[0], p[1], p[2]));
[...]
/boost-1.76.0/include/boost/geometry/geometries/point.hpp:173:107: error: no type named 'type' in 'struct std::enable_if<false, int>'
[...]
````

If I put the initialization of point_t in a constexpr lambda function, that can be evaluated at compile time, I can work around this issue:
````
diff --git a/source/grid/grid_tools.cc b/source/grid/grid_tools.cc
index fcbd2772d1..2cd4da7f03 100644
--- a/source/grid/grid_tools.cc
+++ b/source/grid/grid_tools.cc
@@ -2787,10 +2787,20 @@ namespace GridTools
       {
         using point_t = boost::geometry::model::
           point<double, spacedim, boost::geometry::cs::cartesian>;
-        const auto pt =
-          spacedim == 1 ?
-            point_t(p[0]) :
-            (spacedim == 2 ? point_t(p[0], p[1]) : point_t(p[0], p[1], p[2]));
+        const auto pt([&] {
+          if constexpr (spacedim == 1)
+            {
+              return point_t(p[0]);
+            }
+          else if constexpr (spacedim == 2)
+            {
+              return point_t(p[0], p[1]);
+            }
+          else
+            {
+              return point_t(p[0], p[1], p[2]);
+            }
+        }());
         if (relevant_cell_bounding_boxes_rtree->qbegin(
               boost::geometry::index::intersects(pt)) ==
             relevant_cell_bounding_boxes_rtree->qend())
````

and with this change deal.II compiles and runs again. I suspect the boost version being the issue, but didn't have time yet to look at it in more detail.

**Comment by peterrum** [comment]
@mathsen Thanks for reporting this! I'll create a patch is a few minutes. Unfortunately, we cannot use `if constexpr` (C++17)...

**Pull request merged**